src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs

```diff
@@ -0,0 +1,422 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Script.Binding;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Scripting;
+using Microsoft.CodeAnalysis.Scripting;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    public class CSharpFunctionInvoker : ScriptFunctionInvokerBase
+    {
+        private const string ScriptClassName = "Submission#0";
+
+        private readonly ScriptHost _host;
+        private readonly DictionaryJsonConverter _dictionaryJsonConverter = new DictionaryJsonConverter();
+        private readonly TraceWriter _fileTraceWriter;
+        private readonly FunctionMetadata _functionMetadata;
+        private readonly BindingMetadata _trigger;
+        private readonly Collection<FunctionBinding> _inputBindings;
+        private readonly Collection<FunctionBinding> _outputBindings;
+        private readonly FileSystemWatcher _fileWatcher;
+        private readonly string _triggerParameterName;
+        private readonly bool _omitInputParameter;
+        private readonly IFunctionEntryPointResolver _functionEntryPointResolver;
+        private MethodInfo _function;
+        private FunctionMetadataResolver _metadataResolver;
+        private readonly FunctionAssemblyLoader _assemblyLoader;
+        private Action _reloadScript;
+        private Action _restorePackages;
+
+        private static readonly string[] _watchedFileTypes = { ".cs", ".csx", ".dll", ".exe" };      
+
+        public CSharpFunctionInvoker(ScriptHost host, BindingMetadata trigger, FunctionMetadata metadata, bool omitInputParameter,
+            Collection<FunctionBinding> inputBindings, Collection<FunctionBinding> outputBindings, 
+            IFunctionEntryPointResolver functionEntryPointResolver, FunctionAssemblyLoader assemblyLoader)
+        {
+            _host = host;
+            _trigger = trigger;
+            _triggerParameterName = trigger.Name;
+            _omitInputParameter = omitInputParameter;
+            _inputBindings = inputBindings;
+            _outputBindings = outputBindings;
+            _functionMetadata = metadata;
+            _functionEntryPointResolver = functionEntryPointResolver;
+            _fileWatcher = InitializeFileWatcher(host);
+            _assemblyLoader = assemblyLoader;
+
+            if (_host.ScriptConfig.FileLoggingEnabled)
```

Comment:
Since this code is repeated now in all 3 invoker classes, please move it to a helper in the base class. Possible signature: TraceWriter CreateTraceWriter(ScriptHostConfig config, string functionName)